README.md
# PopIt API

[![Build Status](https://travis-ci.org/mysociety/popit-api.png?branch=master)](https://travis-ci.org/mysociety/popit-api)
[![Dependency Status](https://gemnasium.com/mysociety/popit-api.png)](https://gemnasium.com/mysociety/popit-api)

The API specific part of the [PopIt](http://popit.mysociety.org/) project.
Provides read and write access to the data and manages where it is stored. Will
be [Popolo](http://popoloproject.com/) compliant.

Can be used as a standalone server, or embedded in a
[Express](http://expressjs.com/) backed website.

## WARNING - Alpha code

This is alpha code and subject to frequent and backwards incompatible change.
Feel free to experiment but do not use in production until this message is
removed.

The exact implementation of the [Popolo](http://popoloproject.com/) standard
used here may not be fully up to date, or may contain discrepencies to the
official one. We ([mySociety](http://www.mysociety.org/)) are using this project
to help develop the standard in the light of requirements from our own projects.
The eventual aim is to be Popolo compliant.

## Installation

``` bash
npm install popit-api
```

## Overview

Example of how to create a simple API server:

``` javascript
var express = require('express'),
    apiApp  = require('popit-api');

// Some minimal config is required
var apiConfig = {
  databaseName: 'some-name',
};

// Create the Express app, mount the PopIt api app at the appropriate path and
// start to listen.
var app = express();
app.use('/api', apiApp(apiConfig));
app.listen(3000);
```

You should then be able to go to http://127.0.0.1:3000/api/persons to list all
people (which will not be any initially as the database is empty).

This app provides a REST interface to an API that lets you store
[Popolo](http://popoloproject.com/) compliant data. That's it.

It does not provide any logging or authentication etc. It is intended that you
will do the app that you use this module in.

## Development setup

There is a test app included that you can use to experiment quickly. These
commands will get you a dev environment set up:

``` bash
git clone https://github.com/mysociety/popit-api.git
cd popit-api
npm install .
npm test
node test-server.js
open http://127.0.0.1:3000/
```

When running the tests or the test-server a database called `test-popit-api-db`
is used.

## Configuration

All configuration is done by passing in the config to the app. Currently you
need to either supply one of these two:

``` javascript
// Specify the database name
{ databaseName: 'name-of-mongodb-to-use' }
```

``` javascript
// Use the hostname to decide the db name
{ storageSelector: 'hostName' }
```

Expect the configuration to change significantly as we work out what we actually
need.

## Validation

All data written to the REST API is validated against the schemas stored in
`schemas/popolo`. Local copies are used rather than fetching over http so that
changes can be easily made and experimented with. Note that these schemas may be
different to the current official ones until the standard is finalised.

## REST actions

The REST API is being kept as simple as possible. We'll be adding features as
required.

Responses are always in JSON, or JSONP if a `callback` query is provided. HTTP
Status codes are used to describe the success or failure. If data is returned it
will be something like:

``` json
{ "result": { "id": "123", "name": "Joe Bloggs" } }
{ "result": [ { "id": "123", "name": "Joe Bloggs" }, ... ] }
```

Errors will also return JSON:

```json
{ "errors": [ "Error message - hopefully helpful...", ... ] }
```

### GET `/api/:collection`

Returns an array of all the documents in that collection (currently there is no
filtering or pagination).

### POST `/api/:collection`

Adds the posted document to the collection, if it validates. Can either specify
an `id` or one will be created. Returns `200` with the document as the body
(with the `id` in it).

### PUT `/api/:collection`

Not implemented, will return `405`. Replacing a whole collection is not
something we think that API users will want to do.

### DELETE `/api/:collection`

Not implemented, will return `405`. Deleting a whole collection is not
something we think that API users will want to do.

### GET `/api/:collection/:id`

Returns the document requested, or `404` if it does not exist.

### POST `/api/:collection/:id`

Not implemented, will return `405`. As the documents can not be treated as
collections adding something to them with a `POST` does not apply.

### PUT `/api/:collection/:id`

Replaces the current document with the one provided, if valid. Returns the document with status `200`.

### DELETE `/api/:collection/:id`

Deletes the document. Returns `204` (even if the document did not exist).


## Deployment

You'll probably want to mount this app in your own app. For an example see
`server.js` in the `mysociety-deploy` branch. This is how we deploy it.


CHANGES.md
# Changes

## v0.0.3

  * Handle invalid `GET` and `DELETE` requests where body is empty but Content-Type is JSON
  * Change response of `POST /api/:collection` and `PUT /api/:collection/:id`

## v0.0.2

  * Tidy up dependencies
  * Structure REST tests and use test fixture
  * More details on '/api' call

## v0.0.1

  * initial release


src/app.js
"use strict";

var util        = require('util'),
    assert      = require('assert'),
    express     = require('express'),
    collections = require('./collections'),
    validate    = require('./validate'),
    Storage     = require('../src/storage'),
    packageJSON = require("../package"),
    _           = require('underscore');

_.str = require('underscore.string');

var storageSelectors = {
  fixedName: {
    selector: function (req, res, next) {
      Storage.connectToDatabase(function (err) {
        if (!err) {
          var databaseName = req.app.get('popitApiOptions').databaseName;
          req.storage = new Storage(databaseName);
        }
        next(err);
      });
    },
    optionsCheck: function (options) {
      assert(options.databaseName, "Missing required option 'databaseName'");
    }
  },
  hostName: {
    selector: function (req, res, next) {
      Storage.connectToDatabase(function (err) {
        if (!err) {
          var host = req.host.replace(/\./g, '-');
          var databaseName = 'popit-api-' + _.str.slugify(host);
          req.storage = new Storage(databaseName);
        }
        next(err);
      });
    },
  }
};

module.exports = function (options) {
  
  // Apply defaults to the options here.
  _.defaults(options, {
    storageSelector: 'fixedName'
  });

  // check that we have all the options that we need
  var storageSelector = storageSelectors[options.storageSelector];
  assert(storageSelector, "Could not load storage selector '"+options.storageSelector+"'");
  if (storageSelector.optionsCheck) {
    storageSelector.optionsCheck(options);
  }

  var app = express();

  // store the options
  app.set('popitApiOptions', options);

  // Clean up requests from tools like slumber that set the Content-Type but no body
  // eg https://github.com/dstufft/slumber/pull/32
  app.use( function (req, res, next) {
    if ( (req.method == "GET" || req.method == "DELETE" ) && req.headers['content-type'] === 'application/json' && !req.body ) {
      delete req.headers['content-type'];
    }
    next();
  });


  app.use(express.bodyParser());

  app.use(storageSelector.selector);

  app.get('/', function (req, res) {
    res.jsonp({
      info: {
        databaseName: req.storage.databaseName,
        version:      packageJSON.version,
      },
    });
  });


  /*
    Check that the collection is in the allowed list.
  */
  app.param('collection', function (req, res, next, collection) {

    if (! collections[collection]) {
      res
        .status(404)
        .jsonp({
          errors: ["collection '" + collection + "' not found"]
        });
    } else {
      next();
    }
  });

  app.get('/:collection', function (req, res, next) {
    var collectionName = req.params.collection;
    req.storage.list(collectionName, function (err, docs) {
      if (err) { return next(err); }
      res.jsonp({ result: docs });
    });
  });

  app.get('/:collection/:id', function (req, res, next) {
    var collectionName = req.params.collection;
    var id             = req.params.id;

    req.storage.retrieve( collectionName, id, function (err, doc) {
      if (err) {
        next(err);
      } else if (doc) {
        res.jsonp({ result: doc });
      } else {
        res
          .status(404)
          .jsonp({
            errors: ["id '" + id + "' not found"]
          });
      }
    });
  });

  app.del('/:collection/:id', function (req, res, next) {
    var collectionName = req.params.collection;
    var id             = req.params.id;

    req.storage.delete( collectionName, id, function (err) {
      if (err) {
        next(err);
      } else {
        res
          .status(204)
          .send('');
      }
    });
  });


  function validateBody (req, res, next) {

    var collectionName = req.params.collection;
    var body = req.body;

    // If there is no id create one
    if (!body.id) {
      body.id = req.params.id || Storage.generateID();
    }

    validate(collectionName, body, function (err, errors) {

      if (err) {
        return next(err);
      } else if (errors.length === 0) {
        return next(null);
      } else {

        var details = _.map(errors, function (error) {
          return util.format(
            "Error '%s' with '%s'.",
            error.message,
            error.schemaUri
          );
        });

        res
          .status(400)
          .send({errors: details});
      }
    });
  }


  app.post('/:collection', validateBody, function (req, res, next) {

    var collectionName = req.params.collection;
    var body = req.body;

    req.storage.store(collectionName, body, function (err, doc) {
      if (err) { return next(err); }
      res
        .status(200)
        .jsonp({ result: doc });
    });

  });


  app.put('/:collection/:id', validateBody, function (req, res, next) {

    var collectionName = req.params.collection;
    var id             = req.params.id;
    var body           = req.body;

    if (id !== body.id) {
      return res
        .status(400)
        .jsonp({
          errors: ["URL id and document id are different"]
        });

    }

    req.storage.store(collectionName, body, function (err, doc) {
      if (err) { return next(err); }
      res
        .status(200)
        .jsonp({ result: doc });
    });

  });


  /*
    Anything else we should 400 for as it is probably an unsupported method.
  */
  app.all('*', function (req, res) {
    res.status(405).jsonp({errors: ["unsupported method"] });
  });

  return app;
};
